### ISSUE TITLE
Implementing Windows Shell Extension with Intercom

### ISSUE THREAD

**Issue by JackMordaunt** [Rantanen/intercom#167]
Hi there, it's not clear if this crate allows one to implement [Windows Shell Extensions](https://docs.microsoft.com/en-us/windows/win32/shell/handlers).

Is this possible? If so, what incantations are required? 

Specifically, I'm looking to implement a [Thumbnail Handler](https://docs.microsoft.com/en-us/windows/win32/shell/thumbnail-providers).

Thanks!

**Comment by Rantanen** [comment]
Technically it should be possible. The tricky issues are modeling the various interfaces as Rust traits: `IThumbnailProvider`, but also `IInitializeWithStream` (or similar).

Defining existing interfaces through Rust-traits has already been done ih https://github.com/Rantanen/intercom/blob/master/intercom-cli/src/embed/setup_configuration.rs#L11-L70 so you can use that as an example if you go for it.

Unfortunately the conversion intercom does between traits and COM-methods isn't documented. The basic is that parameters are included in the order they are defined with `String`-parameters converted to `BSTR`. `ComResult<Foo, Bar, Baz>` return type will result in `HRESULT` return type and addition of `Foo* foo, Bar* bar, Baz* baz` output parameters as the last parameters of the method.

In short:
```rust
Rust: fn get_display_name(&self, lcid: u32) -> ComResult<String>
```
```C++
Com:  HRESULT GetDisplayName(uint32_t lcid, BSTR* output)
```

Any `[repr(C)]`-compatible type should be able to be used as is as well. This includes `*mut` pointers, etc. to get around irregular IN/OUT parameter order etc.

The other tricky bit is registering the final dll as intercom doesn't currently implement the facilities for regsvr32. I might find time to implement that at some point. It's been on my list of to-do things.

**Comment by JackMordaunt** [comment]
Thank you, I'll investigate further!

**Comment by Rantanen** [comment]
I got curious about this use case so I made a quick test for it yesterday. Short of needing to newtype all winapi types to be able to derive Intercom traits on them, the implementation was fairly straight forward.

https://github.com/Rantanen/intercom/tree/master/samples/thumbnail_provider

**Comment by JackMordaunt** [comment]
If you don't mind, I'm having some troubles with the sample code. 
I run all the steps "successfully": compile, register server, apply regedit associations.

The description of the file (`.icns` in my case) changes as per the registry value, eg "Rust.Example", but I don't get any thumbnail. 

I added a log to `get_thumbnail()`, and it never appears in the logfile, meaning it never gets called. However, `on_load` does gets called so the server is for sure loaded. 

Furthermore, I kill and restart explorer.exe but this hasn't helped. 

Tangentially, where does the GUID in `association.reg` come from? `{61A6080E-0E9F-3CEF-50A7-622D16145B78}`.

**Comment by Rantanen** [comment]
If you renamed the class or changed the crate name, the CLSID will have changed. You can hard code the CLSID of your IThumbnailProvider implementation in the `com_class` attribute to avoid its auto generation.

```
#[com_class(clsid = "SOMEGUID-0000-0000-0000-000000000000")]
struct ThumbnailProvider;
```

That is exactly what the `{61A6080E-0E9F-3CEF-50A7-622D16145878}` GUID is. By default Intercom auto-generates it by hashing the crate name and type name. If you don't specify a CLSID in `[com_class]` Intercom generates one for you based on the library and type name. This works in many cases as the CLSID isn't often needed - here it is though. I probably should have written the sample with a hard coded CLSID for clarity... :)

**Comment by JackMordaunt** [comment]
Yes that was it, thanks!
